Foundation/NSData.swift

```diff
@@ -465,6 +484,22 @@ open class NSData : NSObject, NSCopying, NSMutableCopying, NSSecureCoding {
         }
     }
 
+#if os(Windows)
+    internal class func write(toHandle handle: HANDLE, path: String? = nil, buf: UnsafeRawPointer, length: Int) throws {
+      var BytesToWrite = length
+      while BytesToWrite > 0 {
+        var BytesWritten: DWORD = 0
+        if WriteFile(handle, buf.advanced(by: length - BytesToWrite), DWORD(BytesToWrite), &BytesWritten, nil) == FALSE {
+          throw _NSErrorWithErrno(Int32(GetLastError()), reading: false, path: path)
+        }
+        if BytesWritten == 0 {
+          throw _NSErrorWithErrno(Int32(GetLastError()), reading: false, path: path)
+        }
+        BytesToWrite -= Int(BytesWritten)
+      }
+    }
+#endif
+
     /// Writes the data object's bytes to the file specified by a given path.
     open func write(toFile path: String, options writeOptionsMask: WritingOptions = []) throws {
```

Comment:
I think it might be better to try and rewrite this whole method using `FileHandle` and eliminate all of the Windows/Unix differences here. It will requires some new `internal` `FileHandle` methods (eg a `getFilePermissions`, `setFilePermissions`, and a throwable version of `synchronizeFile`) but I think it should be possible.
`makeTemporaryFile` could return a `FileHandle` instead of an fd as well. What do you think?

The helper methods could probably be reused in `FileManager` later on any

Reply:
I like that idea, if that is acceptable as per @parkera, @phausler, or @millenomi.  The problem is that this is a public interface, no?

Reply:
I just meant change the function body. Would something like this work (untested, and you will need to add an internal `FileHandle.setFilePermissions(mode)` method:

```
open func write(toFile path: String, options writeOptionsMask: WritingOptions = []) throws {
        let fm = FileManager.default

        let fh: FileHandle
        let useAuxiliaryFile = writeOptionsMask.contains(.atomic)
        var auxFilePath: String? = nil
        var mode: mode_t? = nil

        if useAuxiliaryFile {
            if let perms = try? fm.attributesOfItem(atPath: path)[.posixPermissions] as? NSNumber, let m = perms {
                mode = mode_t(m.uintValue)
            }

            let (newFD, path) = try self.makeTemporaryFile(inDirectory: path._nsObject.deletingLastPathComponent)
            fh = FileHandle(fileDescriptor: newFD, closeOnDealloc: true)
            auxFilePath = path
        } else {
            var flags = O_WRONLY | O_CREAT | O_TRUNC
            if writeOptionsMask.contains(.withoutOverwriting) {
                flags |= O_EXCL
            }

            guard let newfh = FileHandle(path: path, flags: flags, createMode: 0o666) else {
                throw _NSErrorWithErrno(errno, reading: false, path: path)
            }
            fh = newfh
        }

        try self.enumerateByteRangesUsingBlockRethrows { (buf, range, stop) in
            if range.length > 0 {
                do {
                    try NSData.write(toFileDescriptor: fh.fileDescriptor, path: path, buf: buf, length: range.length)
                    fh.synchronizeFile()
                } catch {
                    if let auxFilePath = auxFilePath {
                        try? fm.removeItem(atPath: auxFilePath)
                    }
                    throw error
                }
                fh.setFilePermissions(mode) // TODO: Add this as an internal Filehandle function 
            }
        }

        if let auxFilePath = auxFilePath {
            try fm._fileSystemRepresentation(withPath: path, { pathFsRep in
                try fm._fileSystemRepresentation(withPath: auxFilePath, { auxFilePathFsRep in
                    if rename(auxFilePathFsRep, pathFsRep) != 0 {
                        try? FileManager.default.removeItem(atPath: auxFilePath)
                        throw _NSErrorWithErrno(errno, reading: false, path: path)
                    }
                    if let mode = mode {
                        chmod(pathFsRep, mode)
                    }
                })
            })
        }
    }
```

Reply:
I can rewrite this in terms of `FileHandle`, but note that there will still be Windows specific paths here as `FileHandle.fileDescriptor` is not available on Windows, since it works in terms of `HANDLE`s.  Additionally, a few of the operations that this is performing would need to be moved around (e.g. `chmod`).  The `makeTemporaryFile` will still require Windows specific handling (`FileHandle` wants a `HANDLE`, not a file descriptor).

Reply:
What about moving the two `internal NSData.write()` functions to be internal methods of `FileHandle` (its the only caller of them anyway, and it would match `_readDataOfLength()`) then you should be able to eliminate the `fh.fileDescriptor`.

Reply:
IMO, that is far more accessible for developers working on `FileHandle`.  I remember looking at this and getting confused why it was split.  @parkera @phausler - wanna chime in?

Reply:
I think the only reason why this was two separate implementations is because they are transliterations from the objc implementations. The implementations for NSData live in CoreFoundation and NSFileHandle lives in Foundation. Plus from a historical case the implementations for NSFileHandle use NSExceptions heavily (whereas NSData does not). Frankly these are all implementation details that probably dont need to be maintained across to SCLF. It makes sense to me that file handle things like writing etc should be handled by FileHandle. @millenomi might have some insight here on which way to favor.

Reply:
Ive done a refactor of the `FileHandle` write paths in #1876